### ISSUE TITLE
[Discussion] [Proposal] Code Generation and Scale

### ISSUE THREAD

**Issue by mstrchrstphr** [sschmid/Entitas-CSharp#136]
_I asked a question in gitter yesterday that seemed to spark some proposals for future work so I'd like to move that conversation here for posterity and further discussion. The question was:_
### Does code generation scale?

For large projects that contain hundred (or thousands) of components the `Entity` and `Pool` classes become very cluttered. Very quickly after starting a project the generated API becomes untrustworthy i.e. using most of the generated API methods in any particular scenario would be a logical error and would raise an exception.

@alerdenisov mentioned that this isn't a problem for pools because they could be separated but unless I'm missing something it appears that all `SingleEntity` methods are generated on the base `Pool` class which means most pools are capable of doing things they shouldn't.
#### Entity Lenses

@mzaks proposed experimenting with "Entity Lenses". This involves changing the code gen to output interfaces like so:

``` c#
interface HasPosition{
  Entity addPosition(int x, int y);
  Entity removePosition();
  Entity replacePosition(int x, int y);
}
interface HasVelocity{
  Entity addVelocity(int x, int y);
  Entity removeVelocity();
  Entity replaceVelocity(int x, int y);
}
```

Which we can manually use to create the following:

``` c#
interface MyEntity : HasPosition, HasVelocity {}
```
#### Improved Interfaces and Code Gen

I propose the following changes:
1) Create a formal interface for `Entity`: `IEntity` (I think this should happen regardless).
2) Use the code generator to create entity interfaces that hold all the component related methods for a particular pool:

``` c#
interface CorePoolEntity : IEntity {
  CorePoolEntity AddVelocity(int x, int y);
  CorePoolEntity AddPosition(int x, int y);
  // etc
}
```

3) Make `Pool` generic on `T where T : IEntity`.

``` c#
public partial class Pool<T> where T : IEntity {
  ...
  public virtual T CreateEntity() {
    ...
  }
}
```

4) Create typed pools in the generated `Pools.cs` file:

``` c#
public static Pool<CorePoolEntity> core {
  get {
     ... 
  } 
}
```

I like this solution because it:
1. It completely solves the API scaling problem for entities (I think).
2. Requires no action after you've generated code. Everything is kept up to date automatically.
3. Its safe because it's generated.

Thoughts?

**Comment by bddckr** [comment]
I like it!
This adds safety on pool methods because one cannot use the wrong components on them.

-

But using the `Has*` interfaces @mzaks proposed it's possible to guarantee the usage of a subset of components in a `System` for example:
```c#
class ExampleSystem : IReactiveSystem {
  private interface MyEntity : IEntity, HasPosition, HasVelocity {}
  
  void Execute(IEntity[] entities) {
    for (MyEntity entity in entities.getSubsetOf(typeof(MyEntity))) {
      myEntity.AddPosition(...);
    }
  }
}
```

`getSubsetOf` would be an extension method on `IEntity[]`. It won't be generic since the type is only used to get its inherited interfaces and check the array for entities that implement them all. AFAIK in C# `class A : IOne, ITwo` is **not** of type `IAggregate` where `interface IAggregate: IOne, ITwo`.
If the system uses only one type of `IEntity` (eg. `IReactiveSystem`) its trigger could maybe be replaced by returning `MyEntity` with a trigger event somehow?

-

Maybe it's possible to tell the compiler about everything:

- which components are available in a pool
- which components an entity must have & in which pool the entity must be

**Comment by bddckr** [comment]
If it isn't possible (or the implementation allows no performant fetching) to have both, I'd vote for the pool safety. Having the compiler help with which components are available in a pool sounds better than my idea above.

Just thinking out loud. :smile:

**Comment by mstrchrstphr** [comment]
I agree that @mzaks solution can be used to add extra safety :+1: . I also agree in that pool safety / entity interface cleanliness is my highest priority.

I just started writing C# a few days ago so my knowledge there isn't strong but I'd like to think we could implement both of these solutions.

**Comment by mstrchrstphr** [comment]
Giving this some more thought I realized this would be a pretty serious departure from the current architecture and would break all existing projects. Since pools are generally passed around as just a `Pool` reference I don't think migration tools would be able to migrate to something like this. 

I have a solution for this:

1. Create an `IPool` interface and move existing Pool interface there.
2. Have `Pool` conform to this interface and keep generating `Pool` as we are (with all single entity methods, etc). This is for backwards compatibility.
3. Create a new generic pool class as described in the initial comment:

```c#
public partial class Pool<T> where T : IEntity { }
```

**Comment by sschmid** [comment]
Closed by #266

**Comment by sschmid** [comment]
> Giving this some more thought I realized this would be a pretty serious departure from the current architecture and would break all existing projects

Oh yes, this was a huge change :)
